templates/tftmpl/tmplfunc/hcat.go

```diff
@@ -0,0 +1,119 @@
+// Code here is copied over from hcat's (https://github.com/hashicorp/hcat)
+// internal dependency package that is used for creating custom template
+// functions for CTS
+
+package tmplfunc
+
+import (
+	"context"
+	"sort"
+	"time"
+
+	consulapi "github.com/hashicorp/consul/api"
+)
+
+// QueryOptions is a list of options to send with the query. These options are
+// client-agnostic, and the dependency determines which, if any, of the options
+// to use.
+type QueryOptions struct {
```

Comment:
A lot of this is replicating the additional hcat layer of query options. Do you know if there is a way it work to use the `consulapi.QueryOptions` directly? would it simplify the changes?

Reply:
Great idea!! I tried it out and was able to replace this `QueryOptions` with the `consulapi.QueryOptions` and that cleaned up all this query option code I copied over 🎉. Thanks for the suggestion!

One thing that I debated for awhile is that `consulapi.QueryOption` doesn't have a merge-type function. I tried writing a merge function initially, but because `consulapi.QueryOption` doesn't support nil fields, it couldn't handle some empty values accurately. For now, I opted to write a custom merge within Fetch() since it was less code and a little more accurate.

🤔 I know we typically follow the pattern of having a merge function. I was thinking we could revisit this if we add more template funcs in the future that need to share a merge-type func. Going to go ahead and merge but happy to make changes if you have concerns after-the-fact.

Reply:
Using the direct API query options is probably the best way, but if using the internal QueryOptions ends up being the best solution here please let me know and we can look into making it public in hashicat so you wouldn't need all the copied code.

Reply:
Thanks @eikenb, will do!